File: xlist.py
# /// script
# requires-python = ">=3.11"
# dependencies = [
#     "flask",
#     "pywebview",
#     "qtpy",
#     "PyQt6",
#     "PyQt6-WebEngine",
# ]
# ///

"""
xlist - Personal checklist app (Hearth app).

Stores lists in SQLite database (xlist.db).

Run:
    [uv run] xlist.py                 # native window
    [uv run] xlist.py --serve [port]  # LAN web access

Developer:  KarmaHelen
Contact:    xlist.primary904@passinbox.com
Support:    https://buymeacoffee.com/karmahelen
"""

import json
import re
import sqlite3
import sys
import threading
from pathlib import Path

BASE_DIR = Path(__file__).resolve().parent

# Add parent directory so Hearth can be imported from root
sys.path.insert(0, str(BASE_DIR.parent))

from hearth import run

CONFIG_FILE = BASE_DIR / "xlist.json"

# App-level config, persisted to xlist.json (the standard Hearth per-app pattern).
# controls_collapsed — saved state of the toolbar's controls drawer (toggled directly).
# check_behavior   — what happens to an item when checked: "stay" (remain in place) or
#                    "sink" (drop below the unchecked items). Display-only; see xlist.js.
DEFAULT_CONFIG = {
    "controls_collapsed": False,
    "check_behavior": "stay",
}


def _load_config():
    if CONFIG_FILE.exists():
        with open(CONFIG_FILE) as f:
            raw = f.read()
        # Tolerate trailing commas if the file was hand-edited.
        raw = re.sub(r",\s*([\]}])", r"\1", raw)
        cfg = json.loads(raw)
        # setdefault each key so configs written by older versions gain new keys.
        for k, v in DEFAULT_CONFIG.items():
            cfg.setdefault(k, v)
        return cfg
    return dict(DEFAULT_CONFIG)


def _save_config(cfg):
    with open(CONFIG_FILE, "w") as f:
        json.dump(cfg, f, indent=2)


class XList:
    def __init__(self):
        self.db_path = BASE_DIR / "xlist.db"
        self.config = _load_config()
        # Serializes JSON writes in case serve-mode handles requests concurrently.
        self._config_lock = threading.Lock()
        self._init_db()

    # -- Config --

    def _save_config_locked(self):
        with self._config_lock:
            _save_config(self.config)

    def get_config(self):
        """Curated config for the frontend (mirrors xstocks' get_config pattern)."""
        return {
            "controls_collapsed": self.config["controls_collapsed"],
            "check_behavior": self.config["check_behavior"],
        }

    def set_controls_collapsed(self, collapsed):
        """Persist the controls-drawer state. Called fire-and-forget on toggle."""
        self.config["controls_collapsed"] = bool(collapsed)
        self._save_config_locked()
        return {"controls_collapsed": self.config["controls_collapsed"]}

    def set_settings(self, check_behavior):
        """Persist settings from the Settings view (Save). Batched like xstocks'
        set_settings so new settings can be added as extra params over time."""
        if check_behavior not in ("stay", "sink"):
            raise ValueError("check_behavior must be 'stay' or 'sink'")
        self.config["check_behavior"] = check_behavior
        self._save_config_locked()
        return {"check_behavior": self.config["check_behavior"]}

    # -- DB --

    def _init_db(self):
        con = self._connect()
        con.execute(
            """
            CREATE TABLE IF NOT EXISTS lists (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                name TEXT NOT NULL UNIQUE,
                position INTEGER
            )
            """
        )
        con.execute(
            "CREATE UNIQUE INDEX IF NOT EXISTS idx_lists_position ON lists(position)"
        )
        # checked is a plain column on the item row, not an append-log: an item is
        # either done or not, right now, for one user — there's no history dimension
        # the way xstocks' quotes have. position preserves insertion order (lists
        # carry meaning in their ordering; alphabetizing would destroy it).
        con.execute(
            """
            CREATE TABLE IF NOT EXISTS list_items (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                list_id INTEGER NOT NULL,
                text TEXT NOT NULL,
                checked INTEGER NOT NULL DEFAULT 0,
                position INTEGER NOT NULL,
                FOREIGN KEY (list_id) REFERENCES lists(id)
            )
            """
        )
        con.commit()
        con.close()

    def _connect(self):
        # New connection per call for thread safety in serve mode.
        con = sqlite3.connect(self.db_path)
        con.row_factory = sqlite3.Row
        return con

    # -- Lists --

    def get_lists(self):
        """Return all lists ordered by position."""
        con = self._connect()
        rows = con.execute(
            "SELECT id, name FROM lists ORDER BY position ASC"
        ).fetchall()
        con.close()
        return [dict(r) for r in rows]

    def add_list(self, name, items=""):
        """Create a list. `items` is a raw textarea string, one item per line.
        Items are optional — an empty list is valid (add items afterward)."""
        name = name.strip()
        if not name:
            raise ValueError("List name is required")

        item_texts = [ln.strip() for ln in items.split("\n") if ln.strip()]

        con = self._connect()
        exists = con.execute(
            "SELECT id FROM lists WHERE name = ?", (name,)
        ).fetchone()
        if exists:
            con.close()
            return {"status": "name_exists"}

        next_pos = con.execute(
            "SELECT COALESCE(MAX(position), 0) + 1 FROM lists"
        ).fetchone()[0]
        cur = con.execute(
            "INSERT INTO lists (name, position) VALUES (?, ?)", (name, next_pos)
        )
        list_id = cur.lastrowid
        # Duplicate item text is allowed — a list item isn't an identifier the way
        # a ticker is, so "Battery" twice is legitimate. position + id distinguish them.
        for i, text in enumerate(item_texts, start=1):
            con.execute(
                "INSERT INTO list_items (list_id, text, checked, position) "
                "VALUES (?, ?, 0, ?)",
                (list_id, text, i),
            )
        con.commit()
        con.close()

        return {
            "status": "ok",
            "list": {"id": list_id, "name": name},
            "lists": self.get_lists(),
            "items": self.get_items(list_id),
        }

    def update_list(self, list_id, name, items=""):
        """Update a list's name and items.

        Reconciles items by text using a greedy positional match: each textarea
        line, in order, consumes the first not-yet-used existing row with that text
        (inheriting its checked-state); unmatched lines become new unchecked items;
        leftover existing rows are deleted. The textarea's line order drives item
        position — editing a list is also a reorder. An empty result is allowed."""
        name = name.strip()
        if not name:
            raise ValueError("List name is required")

        item_texts = [ln.strip() for ln in items.split("\n") if ln.strip()]

        con = self._connect()
        # Name conflict with a *different* list
        exists = con.execute(
            "SELECT id FROM lists WHERE name = ? AND id != ?", (name, list_id)
        ).fetchone()
        if exists:
            con.close()
            return {"status": "name_exists"}

        # Existing rows in position order — the pool we match against.
        existing = [
            dict(r)
            for r in con.execute(
                "SELECT id, text, checked FROM list_items "
                "WHERE list_id = ? ORDER BY position ASC, id ASC",
                (list_id,),
            ).fetchall()
        ]
        used = [False] * len(existing)

        # Greedy positional match. ops records the final order as ("keep", id) or
        # ("new", text), so positions can be assigned 1..N in textarea order.
        ops = []
        for text in item_texts:
            matched = False
            for i, row in enumerate(existing):
                if not used[i] and row["text"] == text:
                    used[i] = True
                    ops.append(("keep", row["id"]))
                    matched = True
                    break
            if not matched:
                ops.append(("new", text))

        leftover_ids = [existing[i]["id"] for i in range(len(existing)) if not used[i]]

        # Apply: rename, drop removed rows, then assign positions in textarea order.
        # list_items.position has no UNIQUE constraint, so direct reassignment is safe.
        con.execute("UPDATE lists SET name = ? WHERE id = ?", (name, list_id))
        if leftover_ids:
            placeholders = ",".join("?" * len(leftover_ids))
            con.execute(
                f"DELETE FROM list_items WHERE id IN ({placeholders})", leftover_ids
            )
        for pos, (kind, val) in enumerate(ops, start=1):
            if kind == "keep":
                con.execute(
                    "UPDATE list_items SET position = ? WHERE id = ?", (pos, val)
                )
            else:  # new
                con.execute(
                    "INSERT INTO list_items (list_id, text, checked, position) "
                    "VALUES (?, ?, 0, ?)",
                    (list_id, val, pos),
                )

        con.commit()
        con.close()

        return {
            "status": "ok",
            "list": {"id": list_id, "name": name},
            "lists": self.get_lists(),
            "items": self.get_items(list_id),
        }

    def delete_list(self, list_id):
        """Delete a list and its items, then renumber remaining positions to stay dense."""
        con = self._connect()
        con.execute("DELETE FROM list_items WHERE list_id = ?", (list_id,))
        con.execute("DELETE FROM lists WHERE id = ?", (list_id,))

        # Renumber remaining lists 1..N in current position order. Two-pass
        # (negate, then set positive) to avoid mid-update UNIQUE(position) conflicts.
        remaining_ids = [
            r["id"]
            for r in con.execute(
                "SELECT id FROM lists ORDER BY position ASC"
            ).fetchall()
        ]
        con.execute("UPDATE lists SET position = -position")
        for new_pos, lid in enumerate(remaining_ids, start=1):
            con.execute("UPDATE lists SET position = ? WHERE id = ?", (new_pos, lid))

        con.commit()
        con.close()
        return {"status": "deleted", "lists": self.get_lists()}

    def reorder_lists(self, ordered_ids):
        """Apply a new list ordering. `ordered_ids` must be exactly the current set
        of list ids — no missing, no extra, no duplicates."""
        if not isinstance(ordered_ids, list):
            raise ValueError("ordered_ids must be a list")

        con = self._connect()
        current_ids = {r["id"] for r in con.execute("SELECT id FROM lists").fetchall()}
        new_ids = set(ordered_ids)

        if len(ordered_ids) != len(new_ids):
            con.close()
            raise ValueError("ordered_ids contains duplicates")
        if new_ids != current_ids:
            con.close()
            raise ValueError("ordered_ids must match current list ids exactly")

        # Two-pass update to satisfy UNIQUE(position): negate all (still unique
        # since positions were unique), then set the new positive values.
        con.execute("UPDATE lists SET position = -position")
        for new_pos, lid in enumerate(ordered_ids, start=1):
            con.execute("UPDATE lists SET position = ? WHERE id = ?", (new_pos, lid))
        con.commit()
        con.close()
        return {"status": "reordered", "lists": self.get_lists()}

    # -- Items --

    def get_items(self, list_id):
        """Return a list's items in position order."""
        con = self._connect()
        rows = con.execute(
            "SELECT id, text, checked, position FROM list_items "
            "WHERE list_id = ? ORDER BY position ASC, id ASC",
            (list_id,),
        ).fetchall()
        con.close()
        return [dict(r) for r in rows]

    def add_item(self, list_id, text):
        """Append one item to an existing list."""
        text = text.strip()
        if not text:
            raise ValueError("Item text is required")
        con = self._connect()
        next_pos = con.execute(
            "SELECT COALESCE(MAX(position), 0) + 1 FROM list_items WHERE list_id = ?",
            (list_id,),
        ).fetchone()[0]
        con.execute(
            "INSERT INTO list_items (list_id, text, checked, position) "
            "VALUES (?, ?, 0, ?)",
            (list_id, text, next_pos),
        )
        con.commit()
        con.close()
        return {"items": self.get_items(list_id)}

    def set_checked(self, item_id, checked):
        """Toggle a single item's done-state. Frontend updates optimistically and
        fires this async, so the return value is informational."""
        val = 1 if checked else 0
        con = self._connect()
        con.execute(
            "UPDATE list_items SET checked = ? WHERE id = ?", (val, item_id)
        )
        con.commit()
        con.close()
        return {"id": item_id, "checked": val}

    def clear_list(self, list_id):
        """Reset every item in the list back to unchecked (reuse the list)."""
        con = self._connect()
        con.execute(
            "UPDATE list_items SET checked = 0 WHERE list_id = ?", (list_id,)
        )
        con.commit()
        con.close()
        return {"items": self.get_items(list_id)}


if __name__ == "__main__":
    run(
        XList(),
        frontend=str(BASE_DIR / "xlist.html"),
        title="xlist",
        port=8082,
        window={
            "width": 720,
            "height": 640,
            "min_size": (480, 400),
            "background_color": "#0f1117",
            "text_select": True,
        },
    )

File: xlist.js
// -- State --
let lists = [];
let activeListId = null;
let items = [];

// savedCollapsed = the persisted preference (xlist.json). viewCollapsed = the
// current view's drawer state, derived from the pref at each view-entry but forced
// open on an empty list. They diverge only on empty lists; toggling updates both.
let savedCollapsed = false;
let viewCollapsed = false;

// "stay" (checked items remain in place) or "sink" (checked items drop below the
// unchecked ones). Display-only — never mutates item position. Loaded from xlist.json.
let checkBehavior = 'stay';

// -- Helpers --
// List names and item text are freeform user input injected into innerHTML, so
// escape them (unlike xstocks' tickers, which are constrained identifiers).
function escapeHtml(s) {
    return String(s)
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#39;');
}

function setStatus(msg) {
    document.getElementById('statusText').textContent = msg;
}

function updateCount() {
    const done = items.filter(i => i.checked).length;
    document.getElementById('statusCount').textContent =
        items.length ? `${done} / ${items.length} checked` : '';
}

function checkIcon(checked) {
    return checked
        ? `<svg class="ico ico-check" viewBox="0 0 24 24" aria-hidden="true"><path d="M5 13l4 4L19 7" fill="none" stroke="currentColor" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round"/></svg>`
        : `<svg class="ico ico-x" viewBox="0 0 24 24" aria-hidden="true"><path d="M6 6l12 12M18 6L6 18" fill="none" stroke="currentColor" stroke-width="2.5" stroke-linecap="round"/></svg>`;
}

// Up-chevron; CSS rotates it 180deg when the controls drawer is collapsed.
function chevronIcon() {
    return `<svg class="ico ico-chevron" viewBox="0 0 24 24" aria-hidden="true"><path d="M6 15l6-6 6 6" fill="none" stroke="currentColor" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round"/></svg>`;
}

// -- Init --
async function init() {
    lists = await app.call('get_lists');
    const cfg = await app.call('get_config');
    savedCollapsed = !!cfg.controls_collapsed;
    checkBehavior = cfg.check_behavior || 'stay';
    if (lists.length === 0) {
        showAddList();
        return;
    }
    activeListId = lists[0].id;
    await loadItems();
}

// -- List selector (now lives on the control row, rebuilt by renderList) --
function listSelectHtml() {
    const options = lists.map(l =>
        `<option value="${l.id}" ${l.id === activeListId ? 'selected' : ''}>${escapeHtml(l.name)}</option>`
    ).join('');
    let sentinels = '<option value="__add__">+ Add List</option>';
    if (lists.length > 1) sentinels += '<option value="__reorder__">Reorder Lists</option>';
    return `<select id="listSelect" class="list-select" onchange="handleListChange(this.value)">${options}${sentinels}</select>`;
}

function handleListChange(value) {
    if (value === '__add__') { showAddList(); return; }
    if (value === '__reorder__') { showReorderLists(); return; }
    activeListId = parseInt(value);
    loadItems();
}

// -- Add a New List --
function showAddList() {
    const el = document.getElementById('content');
    el.innerHTML = `
        <div class="setup">
            <h2>Add a New List</h2>
            <input type="text" id="listNameInput" placeholder="List name (e.g. Camping Gear)"
                   onkeydown="if(event.key==='Enter') document.getElementById('listItemsInput').focus()">
            <textarea id="listItemsInput" class="setup-textarea"
                      placeholder="One item per line:&#10;Tent&#10;Sleeping Bag&#10;Backpack"></textarea>
            <div class="form-actions">
                <button class="btn primary" onclick="submitAddList()">Create List</button>
                ${lists.length > 0 ? '<button class="btn" onclick="returnToActiveList()">Cancel</button>' : ''}
            </div>
        </div>
    `;
    document.getElementById('listNameInput').focus();
}

async function submitAddList() {
    const name = document.getElementById('listNameInput').value.trim();
    const itemsText = document.getElementById('listItemsInput').value;
    if (!name) { setStatus('List name is required'); return; }

    setStatus('Creating list...');
    try {
        const data = await app.call('add_list', { name, items: itemsText });
        if (data.status === 'name_exists') {
            setStatus('A list with that name already exists');
            return;
        }
        lists = data.lists;
        activeListId = data.list.id;
        items = data.items;
        viewCollapsed = items.length > 0 ? savedCollapsed : false;
        renderList();
        setStatus('Ready');
        updateCount();
    } catch (e) {
        setStatus('Error: ' + e.message);
    }
}

// Shared exit path from the Add / Edit / Reorder / Settings views back to the active
// list. Clears the reorder guard, then reloads the active list (rebuilding dropdown +
// controls + items). Falls back to the Add form when there are no lists yet.
function returnToActiveList() {
    inReorderView = false;
    if (lists.length === 0 || activeListId == null) {
        showAddList();
        return;
    }
    loadItems();
}

// -- Settings (cogwheel) --
function showSettings() {
    const el = document.getElementById('content');
    el.innerHTML = `
        <div class="setup">
            <h2>Settings</h2>
            <div class="settings-group">
                <div class="lbl">When a list item is checked</div>
                <label class="opt"><input type="radio" name="checkBehavior" value="stay" ${checkBehavior === 'stay' ? 'checked' : ''}> Checked items remain at current position</label>
                <label class="opt"><input type="radio" name="checkBehavior" value="sink" ${checkBehavior === 'sink' ? 'checked' : ''}> Checked items go to the bottom of the list</label>
            </div>
            <div class="form-actions">
                <button class="btn primary" onclick="submitSettings()">Save</button>
                <button class="btn" onclick="returnToActiveList()">Cancel</button>
            </div>
        </div>
    `;
}

async function submitSettings() {
    const sel = document.querySelector('input[name="checkBehavior"]:checked');
    const value = sel ? sel.value : checkBehavior;
    setStatus('Saving settings...');
    try {
        const data = await app.call('set_settings', { check_behavior: value });
        checkBehavior = data.check_behavior;
        // Re-render the list under the new sort, then report saved.
        if (lists.length === 0 || activeListId == null) {
            showAddList();
        } else {
            await loadItems();
        }
        setStatus('Settings saved');
    } catch (e) {
        setStatus('Error: ' + e.message);
    }
}

// -- Edit a List (rename + items + delete) --
function showEditList() {
    const list = lists.find(l => l.id === activeListId);
    if (!list) return;
    const el = document.getElementById('content');
    // Prefill in position order (the order shown) — not sorted. Editing the
    // textarea also reorders, since line order drives item position on submit.
    const itemsText = items.map(i => i.text).join('\n');
    el.innerHTML = `
        <div class="setup">
            <h2>Edit ${escapeHtml(list.name)}</h2>
            <input type="text" id="listNameInput" value="${escapeHtml(list.name)}" placeholder="List name"
                   onkeydown="if(event.key==='Enter') document.getElementById('listItemsInput').focus()">
            <textarea id="listItemsInput" class="setup-textarea"
                      placeholder="One item per line">${escapeHtml(itemsText)}</textarea>
            <div class="form-actions">
                <button class="btn primary" onclick="submitEditList()">Update</button>
                <button class="btn" onclick="returnToActiveList()">Cancel</button>
                <button class="btn danger" onclick="deleteList()">Delete</button>
            </div>
        </div>
    `;
    document.getElementById('listNameInput').focus();
}

async function submitEditList() {
    const name = document.getElementById('listNameInput').value.trim();
    const itemsText = document.getElementById('listItemsInput').value;
    if (!name) { setStatus('List name is required'); return; }

    setStatus('Updating list...');
    try {
        const data = await app.call('update_list', { list_id: activeListId, name, items: itemsText });
        if (data.status === 'name_exists') {
            setStatus('A list with that name already exists');
            return;
        }
        lists = data.lists;
        activeListId = data.list.id;
        items = data.items;
        viewCollapsed = items.length > 0 ? savedCollapsed : false;
        renderList();
        setStatus('Ready');
        updateCount();
    } catch (e) {
        setStatus('Error: ' + e.message);
    }
}

async function deleteList() {
    const list = lists.find(l => l.id === activeListId);
    if (!list) return;
    if (!confirm(`Delete the list "${list.name}" and all its items?`)) return;

    setStatus('Deleting list...');
    try {
        const data = await app.call('delete_list', { list_id: activeListId });
        lists = data.lists;
        if (lists.length > 0) {
            activeListId = lists[0].id;
            await loadItems();
        } else {
            activeListId = null;
            items = [];
            showAddList();
            setStatus('Ready');
            document.getElementById('statusCount').textContent = '';
        }
    } catch (e) {
        setStatus('Error: ' + e.message);
    }
}

// -- Reorder Lists --
// reorderState is a working copy mutated by the user; nothing hits the DB until Update.
// inReorderView guards against a deferred blur-render firing after the user has
// already navigated away (Update/Cancel) — see reorderCommitInput.
let reorderState = [];
let inReorderView = false;

function showReorderLists() {
    inReorderView = true;
    reorderState = lists.map(l => ({ id: l.id, name: l.name }));
    renderReorderView();
}

function renderReorderView() {
    if (!inReorderView) return;
    const el = document.getElementById('content');
    const rows = reorderState.map((l, idx) => `
        <div class="reorder-row">
            <button class="btn reorder-btn" onclick="reorderMoveUp(${l.id})" title="Move up">↑</button>
            <button class="btn reorder-btn" onclick="reorderMoveDown(${l.id})" title="Move down">↓</button>
            <input type="text" class="reorder-pos" value="${idx + 1}" data-id="${l.id}"
                   onkeydown="if(event.key==='Enter'){event.preventDefault();reorderCommitInput(this,false);}"
                   onblur="reorderCommitInput(this,true)">
            <span class="reorder-name">${escapeHtml(l.name)}</span>
        </div>
    `).join('');
    el.innerHTML = `
        <div class="setup">
            <h2>Reorder Lists</h2>
            <div class="reorder-list">${rows}</div>
            <div class="form-actions">
                <button class="btn primary" onclick="submitReorderLists()">Update</button>
                <button class="btn" onclick="returnToActiveList()">Cancel</button>
            </div>
        </div>
    `;
}

function reorderMoveUp(id) {
    const idx = reorderState.findIndex(l => l.id === id);
    if (idx === -1) return;
    if (idx === 0) {
        const [item] = reorderState.splice(0, 1);
        reorderState.push(item);              // wrap: top -> bottom
    } else {
        [reorderState[idx - 1], reorderState[idx]] = [reorderState[idx], reorderState[idx - 1]];
    }
    renderReorderView();
}

function reorderMoveDown(id) {
    const idx = reorderState.findIndex(l => l.id === id);
    if (idx === -1) return;
    const n = reorderState.length;
    if (idx === n - 1) {
        const [item] = reorderState.splice(n - 1, 1);
        reorderState.unshift(item);           // wrap: bottom -> top
    } else {
        [reorderState[idx], reorderState[idx + 1]] = [reorderState[idx + 1], reorderState[idx]];
    }
    renderReorderView();
}

function reorderCommitInput(inputEl, fromBlur) {
    const id = parseInt(inputEl.dataset.id, 10);
    const oldIdx = reorderState.findIndex(l => l.id === id);
    if (oldIdx === -1) return;
    const raw = inputEl.value.trim();
    const n = reorderState.length;

    // Valid only if a clean integer in range 1..n and actually different.
    const valid = /^\d+$/.test(raw) && +raw >= 1 && +raw <= n && +raw !== oldIdx + 1;
    if (valid) {
        const [item] = reorderState.splice(oldIdx, 1);
        reorderState.splice(+raw - 1, 0, item);
    }

    // From blur: defer so a pending click (Move buttons, Update, Cancel) fires first
    // on the still-existing DOM. From Enter: render immediately.
    if (fromBlur) {
        setTimeout(renderReorderView, 0);
    } else {
        renderReorderView();
    }
}

async function submitReorderLists() {
    inReorderView = false;
    try {
        const data = await app.call('reorder_lists', { ordered_ids: reorderState.map(l => l.id) });
        lists = data.lists;
        await loadItems();          // rebuilds the reordered dropdown + items
        setStatus('Lists reordered');
    } catch (e) {
        inReorderView = true;       // stay in the view so the user can retry
        setStatus('Error: ' + e.message);
    }
}

// -- Load + render the active list --
async function loadItems() {
    setStatus('Loading...');
    items = await app.call('get_items', { list_id: activeListId });
    // View-entry: derive the drawer state from the saved pref, but force it open on
    // an empty list (you'd otherwise have no visible way to add the first item).
    viewCollapsed = items.length > 0 ? savedCollapsed : false;
    renderList();
    setStatus('Ready');
    updateCount();
}

// Display order is derived, never stored: "stay" shows items in authored position
// order; "sink" partitions unchecked-then-checked, each group keeping position order.
function displayItems() {
    if (checkBehavior !== 'sink') return items;
    return items.filter(i => !i.checked).concat(items.filter(i => i.checked));
}

function itemRowsHtml() {
    return displayItems().map(it => `
        <div class="item-row ${it.checked ? 'checked' : ''}" data-id="${it.id}" onclick="toggleItem(${it.id})">
            <span class="item-check">${checkIcon(it.checked)}</span>
            <span class="item-text">${escapeHtml(it.text)}</span>
        </div>`).join('');
}

function renderList() {
    const el = document.getElementById('content');
    const hasItems = items.length > 0;
    const collapsedClass = viewCollapsed ? ' collapsed' : '';

    // Control row (always visible): list selector + the controls toggle.
    // Navigation stays put; management (the drawer below) folds away.
    const controlRow = `
        <div class="control-row">
            ${listSelectHtml()}
            ${hasItems ? `<button class="ctrl-toggle" aria-label="Show or hide controls" onclick="toggleControls()">${chevronIcon()}</button>` : ''}
        </div>`;

    // Collapsible drawer: the management controls.
    const drawer = `
        <div class="drawer"><div class="drawer-inner">
            <div class="control-actions">
                <input type="text" id="addItemInput" class="add-item-input" placeholder="Add item..."
                       onkeydown="if(event.key==='Enter') addItem()">
                <button class="btn" onclick="addItem()">Add</button>
                <button class="btn" onclick="showEditList()">Edit</button>
                ${hasItems ? '<button class="btn" onclick="clearList()">Clear</button>' : ''}
            </div>
        </div></div>`;

    const body = hasItems
        ? `<div class="item-list">${itemRowsHtml()}</div>`
        : `<div class="empty-note">This list is empty.<br>Add an item above to get started.</div>`;

    el.innerHTML =
        `<div class="list-controls${collapsedClass}" id="listControls">${controlRow}${drawer}</div>${body}`;
}

// Toggle the controls drawer. Flips a class on the existing wrapper (so CSS animates
// the collapse) rather than re-rendering, and persists the preference fire-and-forget.
function toggleControls() {
    viewCollapsed = !viewCollapsed;
    savedCollapsed = viewCollapsed;
    const wrap = document.getElementById('listControls');
    if (wrap) wrap.classList.toggle('collapsed', viewCollapsed);
    if (!viewCollapsed) {
        const inp = document.getElementById('addItemInput');
        if (inp) inp.focus();
    }
    app.call('set_controls_collapsed', { collapsed: viewCollapsed })
        .catch(() => setStatus('Could not save view state'));
}

// -- Item interactions --

// Optimistic: update the DOM immediately, persist async, revert on failure.
// "stay": targeted icon flip, no movement. "sink": re-sort the item list with a
// FLIP glide so the toggled item animates to its new position.
async function toggleItem(id) {
    const it = items.find(i => i.id === id);
    if (!it) return;

    const newChecked = it.checked ? 0 : 1;
    applyToggle(it, newChecked);
    updateCount();

    try {
        await app.call('set_checked', { item_id: id, checked: newChecked });
    } catch (e) {
        applyToggle(it, newChecked ? 0 : 1);   // revert
        updateCount();
        setStatus('Error saving: ' + e.message);
    }
}

function applyToggle(it, checked) {
    it.checked = checked;
    if (checkBehavior === 'sink') {
        flipReorder();
    } else {
        applyRowState(it.id, checked);
    }
}

function applyRowState(id, checked) {
    const row = document.querySelector(`.item-row[data-id="${id}"]`);
    if (!row) return;
    row.classList.toggle('checked', !!checked);
    const cell = row.querySelector('.item-check');
    if (cell) cell.innerHTML = checkIcon(checked);
}

// FLIP: record row positions (First), re-render the item list in the new sorted order
// (Last), then translate each row from its old spot and transition to zero (Invert/Play).
// Only the .item-list is rebuilt, so the add-item input's text and focus are preserved.
function flipReorder() {
    const listEl = document.querySelector('.item-list');
    if (!listEl) return;

    const first = {};
    listEl.querySelectorAll('.item-row').forEach(r => {
        first[r.dataset.id] = r.getBoundingClientRect().top;
    });

    listEl.innerHTML = itemRowsHtml();

    listEl.querySelectorAll('.item-row').forEach(r => {
        const prevTop = first[r.dataset.id];
        if (prevTop == null) return;
        const delta = prevTop - r.getBoundingClientRect().top;
        if (!delta) return;
        r.style.transition = 'transform 0s';
        r.style.transform = `translateY(${delta}px)`;
        requestAnimationFrame(() => {
            r.style.transition = 'transform 0.28s cubic-bezier(.2,.7,.3,1)';
            r.style.transform = '';
        });
        r.addEventListener('transitionend', function te() {
            r.style.transition = '';   // restore the CSS hover transition afterward
            r.removeEventListener('transitionend', te);
        });
    });
}

// add/clear re-render the list, which reads the unchanged viewCollapsed — so the
// drawer keeps whatever state it's in (these are only reachable while it's open).
async function addItem() {
    const input = document.getElementById('addItemInput');
    const text = input.value.trim();
    if (!text) return;

    setStatus('Adding item...');
    try {
        const data = await app.call('add_item', { list_id: activeListId, text });
        items = data.items;
        renderList();
        document.getElementById('addItemInput').focus();
        setStatus('Ready');
        updateCount();
    } catch (e) {
        setStatus('Error: ' + e.message);
    }
}

async function clearList() {
    if (!confirm('Clear all checkmarks on this list?')) return;

    setStatus('Clearing...');
    try {
        const data = await app.call('clear_list', { list_id: activeListId });
        items = data.items;
        renderList();
        setStatus('Ready');
        updateCount();
    } catch (e) {
        setStatus('Error: ' + e.message);
    }
}

document.addEventListener('DOMContentLoaded', init);
